cirq/linalg/decompositions.py

```diff
@@ -149,6 +149,50 @@ def _perp_eigendecompose(matrix: np.ndarray,
     return vals, vecs
 
 
+def closest_unitary(matrix: np.ndarray) -> np.ndarray:
+    """Returns the closest unitary to the given matrix in Frobenius norm.
+
+    The closest unitary is the orthogonal factor in the polar decomposition
+    of a matrix. For all A with singular value decomposition
+    $A = V S W^H$ and left polar decomposition $A = UJ$,
+    where V, W and are U unitary, S is the diagonal matrix of A's
+    singular values, J is positive semi-definite, we can see that
+    $J = W S W^H$ and the orthogonal component is $U = VW^H$.
+
+    This case:
+    $|| A - U || <= || A - M ||$ for all unitary M.
+
+    For proof see Theorem 1 in:
+    https://www.ams.org/journals/proc/1955-006-01/S0002-9939-1955-0067841-7/S0002-9939-1955-0067841-7.pdf
+    """
+    V, __, Wd = np.linalg.svd(matrix)
+    return V @ Wd
+
+
+def unitary_eig(matrix: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+    """Returns a guaranteed unitary diagonalization of a unitary matrix
+    up to 1e-14.
```

Comment:
Is this bound still true when the given matrix has `dtype=np.float32`?

Reply:
ignore this - I changed direction based on @viathor's suggestions.